From 93d07a8da1e406402b222d073c55f17b8623f5d6 Mon Sep 17 00:00:00 2001
From: Grzegorz Jaszczyk <jaz@semihalf.com>
Date: Wed, 18 Mar 2020 23:10:18 +0100
Subject: [PATCH 1123/1239] sar: mvebu: ap806: use smc calls for dfx
 register-set access

This change will be needed after dfx register set will be secured by the
trusted firmware.

Change-Id: I355faa4afb0050713e7db12e9d122d9466d58348
Signed-off-by: Grzegorz Jaszczyk <jaz@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/25183
Reviewed-by: Kostya Porotchkin <kostap@marvell.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
---
 drivers/misc/mvebu_sar/ap806_sar.c | 24 +++++++++++++++++++-----
 1 file changed, 19 insertions(+), 5 deletions(-)

diff --git a/drivers/misc/mvebu_sar/ap806_sar.c b/drivers/misc/mvebu_sar/ap806_sar.c
index a19de16778..3cda3d3b63 100644
--- a/drivers/misc/mvebu_sar/ap806_sar.c
+++ b/drivers/misc/mvebu_sar/ap806_sar.c
@@ -148,11 +148,17 @@ static const u32 pll_freq_tbl[16][4] = {
 
 static u32 sar_get_clock_freq_mode(struct udevice *dev)
 {
-	u32 i;
+	u32 i, clock_freq;
 	struct dm_sar_pdata *priv = dev_get_priv(dev);
+	int ret;
+
+	ret = mvebu_dfx_sread(&clock_freq, (uintptr_t)priv->sar_base);
+	if (ret != SMCCC_RET_SUCCESS) /* try legacy read */
+		clock_freq = (readl(priv->sar_base));
+
+	clock_freq = (clock_freq & SAR_CLOCK_FREQ_MODE_MASK) >>
+		      SAR_CLOCK_FREQ_MODE_OFFSET;
 
-	u32 clock_freq = (readl(priv->sar_base) & SAR_CLOCK_FREQ_MODE_MASK) >>
-			  SAR_CLOCK_FREQ_MODE_OFFSET;
 	for (i = 0; i < 16; i++) {
 		if (pll_freq_tbl[i][3] == clock_freq)
 			return i;
@@ -191,8 +197,12 @@ static int ap806_sar_dump(struct udevice *dev)
 	u32 reg, val;
 	struct sar_info *sar;
 	struct dm_sar_pdata *priv = dev_get_priv(dev);
+	int ret;
+
+	ret = mvebu_dfx_sread(&reg, (uintptr_t)priv->sar_base);
+	if (ret != SMCCC_RET_SUCCESS) /* try legacy */
+		reg = readl(priv->sar_base);
 
-	reg = readl(priv->sar_base);
 	printf("AP806 SAR register 0 [0x%08x]:\n", reg);
 	printf("----------------------------------\n");
 	sar = ap806_sar_0;
@@ -202,7 +212,11 @@ static int ap806_sar_dump(struct udevice *dev)
 		sar++;
 	}
 
-	reg = readl(priv->sar_base + AP806_SAR_1_REG);
+	ret = mvebu_dfx_sread(&reg,
+			      (uintptr_t)(priv->sar_base + AP806_SAR_1_REG));
+	if (ret != SMCCC_RET_SUCCESS) /* try legacy */
+		reg = readl(priv->sar_base + AP806_SAR_1_REG);
+
 	printf("\nAP806 SAR register 1 [0x%08x]:\n", reg);
 	printf("----------------------------------\n");
 	sar = ap806_sar_1;
-- 
2.29.0

